File: src/AVLTree.cpp
#include <algorithm>

// ===============================
// AVL NODE DEFINITION
// ===============================
struct AVLNode {
    int key;
    int height;
    AVLNode* left;
    AVLNode* right;

    AVLNode(int k) {
        key = k;
        height = 1;
        left = nullptr;
        right = nullptr;
    }
};

// ===============================
// AVL TREE CLASS
// ===============================
class AVLTree {

private:
    AVLNode* root;

    // -------- HEIGHT --------
    int height(AVLNode* node) {
        return (node == nullptr) ? 0 : node->height;
    }

    // -------- BALANCE FACTOR --------
    int getBalance(AVLNode* node) {
        if (node == nullptr) return 0;
        return height(node->left) - height(node->right);
    }

    // -------- RIGHT ROTATION --------
    AVLNode* rightRotate(AVLNode* y) {
        AVLNode* x = y->left;
        AVLNode* T2 = x->right;

        x->right = y;
        y->left = T2;

        y->height = 1 + std::max(height(y->left), height(y->right));
        x->height = 1 + std::max(height(x->left), height(x->right));

        return x;
    }

    // -------- LEFT ROTATION --------
    AVLNode* leftRotate(AVLNode* x) {
        AVLNode* y = x->right;
        AVLNode* T2 = y->left;

        y->left = x;
        x->right = T2;

        x->height = 1 + std::max(height(x->left), height(x->right));
        y->height = 1 + std::max(height(y->left), height(y->right));

        return y;
    }

    // -------- INSERT --------
    AVLNode* insert(AVLNode* node, int key) {

        // Normal BST insertion
        if (node == nullptr)
            return new AVLNode(key);

        if (key < node->key)
            node->left = insert(node->left, key);
        else if (key > node->key)
            node->right = insert(node->right, key);
        else
            return node; // duplicate keys not allowed

        // Update height
        node->height = 1 + std::max(height(node->left), height(node->right));

        // Balance factor
        int balance = getBalance(node);

        // ---- ROTATION CASES ----

        // Left Left
        if (balance > 1 && key < node->left->key)
            return rightRotate(node);

        // Right Right
        if (balance < -1 && key > node->right->key)
            return leftRotate(node);

        // Left Right
        if (balance > 1 && key > node->left->key) {
            node->left = leftRotate(node->left);
            return rightRotate(node);
        }

        // Right Left
        if (balance < -1 && key < node->right->key) {
            node->right = rightRotate(node->right);
            return leftRotate(node);
        }

        return node;
    }

    // -------- SEARCH (RECURSIVE) --------
    AVLNode* search(AVLNode* node, int key) {
        if (node == nullptr || node->key == key)
            return node;

        if (key < node->key)
            return search(node->left, key);

        return search(node->right, key);
    }

public:
    // -------- CONSTRUCTOR --------
    AVLTree() {
        root = nullptr;
    }

    // -------- INSERT PUBLIC --------
    void insertKey(int key) {
        root = insert(root, key);
    }

    // -------- SEARCH PUBLIC --------
    bool searchKey(int key) {
        return search(root, key) != nullptr;
    }
};

// ===============================
// MAIN (NO I/O — LOGIC ONLY)
// ===============================
int main() {

    AVLTree tree;

    // Sample insertions
    tree.insertKey(50);
    tree.insertKey(30);
    tree.insertKey(70);
    tree.insertKey(20);
    tree.insertKey(40);
    tree.insertKey(60);
    tree.insertKey(80);

    // Sample searches (results stored, not printed)
    bool found1 = tree.searchKey(40); // true
    bool found2 = tree.searchKey(25); // false

    return 0;
}

File: java-ui/src/AVLTree.java
// AVLTree.java
public class AVLTree {
    private AVLNode root;

    public void insert(String key, VehicleRecord value) {
        root = insertRec(root, key, value);
    }

    private AVLNode insertRec(AVLNode node, String key, VehicleRecord value) {
        if (node == null) return new AVLNode(key, value);
        int cmp = key.compareTo(node.key);
        if (cmp < 0) node.left = insertRec(node.left, key, value);
        else if (cmp > 0) node.right = insertRec(node.right, key, value);
        else node.value = value; // update
        node.height = 1 + Math.max(height(node.left), height(node.right));
        return balance(node);
    }

    public VehicleRecord find(String key) {
        AVLNode cur = root;
        while (cur != null) {
            int cmp = key.compareTo(cur.key);
            if (cmp == 0) return cur.value;
            if (cmp < 0) cur = cur.left; else cur = cur.right;
        }
        return null;
    }

    private int height(AVLNode n) { return n == null ? 0 : n.height; }

    private int bf(AVLNode n) { return n == null ? 0 : height(n.left) - height(n.right); }

    private AVLNode rotateRight(AVLNode y) {
        AVLNode x = y.left;
        AVLNode T2 = x.right;
        x.right = y; y.left = T2;
        y.height = 1 + Math.max(height(y.left), height(y.right));
        x.height = 1 + Math.max(height(x.left), height(x.right));
        return x;
    }

    private AVLNode rotateLeft(AVLNode x) {
        AVLNode y = x.right;
        AVLNode T2 = y.left;
        y.left = x; x.right = T2;
        x.height = 1 + Math.max(height(x.left), height(x.right));
        y.height = 1 + Math.max(height(y.left), height(y.right));
        return y;
    }

    private AVLNode balance(AVLNode node) {
        int balance = bf(node);
        if (balance > 1 && bf(node.left) >= 0) return rotateRight(node);
        if (balance > 1 && bf(node.left) < 0) {
            node.left = rotateLeft(node.left);
            return rotateRight(node);
        }
        if (balance < -1 && bf(node.right) <= 0) return rotateLeft(node);
        if (balance < -1 && bf(node.right) > 0) {
            node.right = rotateRight(node.right);
            return rotateLeft(node);
        }
        return node;
    }
}
